File: solvers/sampler.py
"""Step sampling / throttling for the on_step callback.

A search can emit millions of events (Algo1 hits ~1.4M nodes on the hardest
benchmark puzzles). ``StepSampler`` sits between the solver and the caller's
``on_step`` and forwards an evenly-spaced subset.

Contract
--------
* ``max_events is None``            -> forward every event.
* ``on_step is None``              -> the solver never builds an emitter and
                                     ``emit`` is a no-op reference (near-zero
                                     cost; see the overhead benchmark).
* otherwise the sampler forwards an event iff the *primary tick* it belongs to
  is on the stride, where::

      stride = max(1, ceil(work_budget / max_events))

  ``work_budget`` is ``max_steps`` when given, else ``DEFAULT_WORK_BUDGET``.
  Sub-events of a sampled tick (e.g. the eliminations triggered by one
  assignment) are all forwarded, so each delivered frame is coherent.
* Lifecycle events (start/phase/solved/stopped) are always forwarded and do
  not count against ``max_events``.
* A hard stop: once ``max_events`` non-lifecycle events have been delivered,
  further non-lifecycle events are dropped. With correct stride sizing this
  is rarely reached; it only bounds the eliminate-heavy tail.

The solver calls ``sampler.tick()`` once per primary step and
``sampler.emit(type, **fields)`` for every event.
"""
import math

from .events import LIFECYCLE

DEFAULT_WORK_BUDGET = 1_000_000


class StepSampler:
    __slots__ = ("_on_step", "_max_events", "_stride", "_tick", "_tick_live",
                 "_delivered", "_stopped")

    def __init__(self, on_step, max_events=None, max_steps=None):
        self._on_step = on_step
        self._max_events = max_events
        budget = max_steps if max_steps else DEFAULT_WORK_BUDGET
        if max_events is None or max_events <= 0:
            self._stride = 1
        else:
            self._stride = max(1, math.ceil(budget / max_events))
        self._tick = 0
        self._tick_live = True          # is the current tick on the stride?
        self._delivered = 0
        self._stopped = False

    # -- called once per primary step (node / iteration / arc revision) -------
    def tick(self):
        self._tick += 1
        self._tick_live = (self._tick % self._stride == 0)
        return self._tick

    @property
    def step(self):
        return self._tick

    @property
    def delivered(self):
        return self._delivered

    # -- called for every event --------------------------------------------
    def emit(self, ev_type, **fields):
        is_life = ev_type in LIFECYCLE
        if not is_life:
            if self._stopped or not self._tick_live:
                return
            if self._max_events is not None and self._delivered >= self._max_events:
                self._stopped = True
                return
        ev = {"type": ev_type, "step": self._tick, **fields}
        if not is_life:
            self._delivered += 1
        self._on_step(ev)

File: web/packages/solver-core/src/sampler.ts
/**
 * Port of solvers/sampler.py::StepSampler.
 *
 * Forwards an evenly-spaced subset of events to `onStep`:
 *   stride = max(1, ceil(workBudget / maxEvents)),  workBudget = maxSteps or 1e6
 * A primary tick is sampled iff `tick % stride === 0`; all sub-events of a
 * sampled tick are forwarded. Lifecycle events always pass and don't count
 * toward maxEvents. Once maxEvents non-lifecycle events are delivered, further
 * non-lifecycle events are dropped.
 */
import { LIFECYCLE_TYPES, type StepEvent } from "./types.js";

const DEFAULT_WORK_BUDGET = 1_000_000;

export class StepSampler {
  private readonly onStep: (ev: StepEvent) => void;
  private readonly maxEvents: number | undefined;
  private readonly stride: number;
  private tickCount = 0;
  private tickLive = true;
  private deliveredCount = 0;
  private stopped = false;

  constructor(
    onStep: (ev: StepEvent) => void,
    maxEvents?: number,
    maxSteps?: number,
  ) {
    this.onStep = onStep;
    this.maxEvents = maxEvents;
    const budget = maxSteps && maxSteps > 0 ? maxSteps : DEFAULT_WORK_BUDGET;
    this.stride =
      maxEvents === undefined || maxEvents <= 0
        ? 1
        : Math.max(1, Math.ceil(budget / maxEvents));
  }

  /** Once per primary step (search node / iteration / arc revision). */
  tick(): number {
    this.tickCount += 1;
    this.tickLive = this.tickCount % this.stride === 0;
    return this.tickCount;
  }

  get step(): number {
    return this.tickCount;
  }
  get delivered(): number {
    return this.deliveredCount;
  }

  /** Build and (maybe) forward an event. `fields` excludes `type` and `step`. */
  emit(type: StepEvent["type"], fields: Record<string, unknown> = {}): void {
    const isLifecycle = LIFECYCLE_TYPES.has(type);
    if (!isLifecycle) {
      if (this.stopped || !this.tickLive) return;
      if (this.maxEvents !== undefined && this.deliveredCount >= this.maxEvents) {
        this.stopped = true;
        return;
      }
    }
    const ev = { type, step: this.tickCount, ...fields } as unknown as StepEvent;
    if (!isLifecycle) this.deliveredCount += 1;
    this.onStep(ev);
  }
}
